### DISCUSSION TITLE
Create span from a trace ID

### DISCUSSION THREAD

**Discussion by irajhedayati** [open-telemetry/opentelemetry-java#4593] [Q&A] [answered]
Hi,
I'm working on a Dataflow job that pulls messages from a Pub/Sub topic, does something, and publishes them to another Pub/Sub topic. The idea is that we generate a trace ID in the source and propagate it throughout the pipeline. Note that there is no REST communication. The only info available is just trace ID and no span ID.

In my job, I'm trying to create a span from the given trace ID. But everything I read so far points me to use the propagators which need both trace ID and span ID. Otherwise, the span context will be invalid and the library generates a new trace ID.

This is what I'm trying to do. In this code, I return a fixed value for trace ID but in reality, the getter extracts that value from the input message.

```java

  private transient Tracer tracer;
  private transient TextMapGetter<I> getter;

  @StartBundle
  public void setup(PipelineOptions pipelineOptions) {
    JobOptions options = pipelineOptions.as(JobOptions.class);
    this.tracer =
        GlobalOpenTelemetry.getTracer(options.getTracingAppName(), options.getTracingAppVersion());
    this.getter =
        new TextMapGetter<>() {
          @Override
          public Iterable<String> keys(I carrier) {return Collections.emptyList();}

          @Override
          public String get(@Nullable I carrier, String key) {return "b15efcf1b2ac2079245e6044ce09b091";}
        };
  }

  @ProcessElement
  public void processElement(@Element I message, OutputReceiver<O> receiver) {
    Context extractedContext =
        GlobalOpenTelemetry.getPropagators()
            .getTextMapPropagator()
            .extract(Context.current(), message, getter);

    try (Scope unused = extractedContext.makeCurrent()) {
      Span span =
          tracer
              .spanBuilder("dataflow job 1")
              .setSpanKind(SpanKind.CONSUMER)
              .setAttribute("thread.id", Thread.currentThread().getId())
              .setAttribute("thread.name", Thread.currentThread().getName())
              .setAttribute("code.function", getOperationName())
              .setAttribute("code.namespace", this.getClass().getName())
              .startSpan();
      try {
        operation(message, receiver);
        span.setStatus(StatusCode.OK);
      } catch (Exception e) {
        span.recordException(e);
        span.setStatus(StatusCode.ERROR, e.getClass().getSimpleName());
        throw e;
      } finally {
        span.end();
      }
    }
  }
```

**Comment by anuraaga** [comment, marked as answer]
Hi @irajhedayati - we'd generally expect the source to create a `PRODUCER` span too, and you'd propagate both the trace ID and span ID as message attributes. Is there a reason you only generate a trace ID at the source?

**Reply by irajhedayati** [reply]
I don't have a reason, to be honest. We are building a pipeline that is composed of multiple components. The input to this pipeline has a trace ID but no span ID. The other components implemented in the Go language, are able to create spans based on only trace ID. But the Java library doesn't support that. I was wondering if there is any reason that these two libraries have different approaches.

**Reply by anuraaga** [reply]
The opentelemetry specification explicitly requires a trace and span ID to both be present to have a valid identifier

https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#isvalid

so Java has implemented this, meaning that both are required when continuing a trace. Do you see opentelemetry-go have different behavior? In that case it might only be loosely implementing this spec.

**Reply by irajhedayati** [reply]
I have the same hunch. I'll find out how the other library is generating spans based on only trace ID and get back to you.

**Reply by irajhedayati** [reply]
Thanks @anuraaga for your input. Based on the specification, the parent should have the span ID.
